zds/notification/models.py

```diff
@@ -0,0 +1,300 @@
+from smtplib import SMTPException
+
+from django.contrib.auth.models import User
+from django.contrib.contenttypes.fields import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
+from django.core.mail import EmailMultiAlternatives
+from django.db import models
+from django.template.loader import render_to_string
+from django.utils.translation import ugettext_lazy as _
+
+from zds import settings
+from zds.forum.models import Topic
+from zds.notification.managers import NotificationManager, SubscriptionManager, TopicFollowedManager
+from zds.utils.misc import convert_camel_to_underscore
+
+
+class Subscription(models.Model):
+    """
+    Model used to register the subscription of a user to a set of notifications (regarding a tutorial, a forum, ...)
+    """
+
+    class Meta:
+        verbose_name = _(u'Abonnement')
+        verbose_name_plural = _(u'Abonnements')
+
+    user = models.ForeignKey(User, related_name='subscriber', db_index=True)
+    pubdate = models.DateTimeField(_(u'Date de création'), auto_now_add=True, db_index=True)
+    is_active = models.BooleanField(_(u'Actif'), default=True, db_index=True)
+    by_email = models.BooleanField(_(u'Recevoir un email'), default=False)
+    content_type = models.ForeignKey(ContentType)
+    object_id = models.PositiveIntegerField()
+    content_object = GenericForeignKey('content_type', 'object_id')
+    last_notification = models.ForeignKey(u'Notification', related_name="last_notification", null=True, default=None)
+
+    def __unicode__(self):
+        return _(u'<Abonnement du membre "{0}" aux notifications pour le {1}, #{2}>')\
+            .format(self.user.username, self.content_type, self.object_id)
+
+    def activate(self):
+        """
+        Activates the subscription if it's inactive. Does nothing otherwise
+        """
+        if not self.is_active:
+            self.is_active = True
+            self.save()
+
+    def activate_email(self):
+        """
+        Activates the notifications by email (and the subscription itself if needed)
+        """
+        if not self.is_active or not self.by_email:
+            self.is_active = True
+            self.by_email = True
+            self.save()
+
+    def deactivate(self):
+        """
+        Deactivate the subscription if it is active. Does nothing otherwise
+        """
+        if self.is_active:
+            self.is_active = False
+            self.save()
+
+    def deactivate_email(self):
+        """
+        Deactivate the email if it is active. Does nothing otherwise
+        """
+        if self.is_active and self.by_email:
+            self.by_email = False
+            self.save()
+
+    def set_last_notification(self, notification):
+        """
+        Replace last_notification by the one given
+        """
+        self.last_notification = notification
+        self.save()
+
+    def send_email(self, notification):
+        """
+        Sends an email notification
```

Comment:
C'est à dire "destiné au développeur" et "pas utile". La fonction peut bien
générer une AssertionError non?

2016-02-25 14:19 GMT+01:00 Gérard Paligot <notifications@github.com>:

> In zds/notification/models.py
> <https://github.com/zestedesavoir/zds-site/pull/3301#discussion_r54092789>
> :
>
> > +        Deactivate the email if it is active. Does nothing otherwise
> > +        """
> > +        if self.is_active and self.by_email:
> > +            self.by_email = False
> > +            self.save()
> > +
> > +    def set_last_notification(self, notification):
> > +        """
> > +        Replace last_notification by the one given
> > +        """
> > +        self.last_notification = notification
> > +        self.save()
> > +
> > +    def send_email(self, notification):
> > +        """
> > +        Sends an email notification
>
> Pas utile.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zestedesavoir/zds-site/pull/3301/files#r54092789>.
>